### PR TITLE
Support half-precision dtypes when calculating model size

### PR DESCRIPTION
**Pull request by jedi2610**
@TylerYep 2 tests from `torchinfo_xl_test.py` are failing for me. can you check if it works for you? cause those 2 tests don't work for me from master as well. hope this fixes the issue. do check out the code and lmk if i have to make any changes. ty

### CODE REVIEW

**Comment by jedi2610** [comment]
@TylerYep nvm didn't know you had github actions on. another thing is i still have to check for `half()` models. how should i do it? should i like create another test case for all `half()` models? (that would result in 2*test cases). am i right?

**Comment by TylerYep** [comment]
I would just add a couple new test cases (don't change the old ones). I would try it using a Linear layer, SingleInputNet, an LSTM, etc

**Comment by jedi2610** [comment]
@TylerYep cool. i have college, so i'll prolly add tests in a day or two.

**Comment by jedi2610** [comment]
@TylerYep all test cases seem to pass fine except when we pass in `input_size` for a `half()` model and there few edge cases where `summary` would fail (especially when running on cpu). take a look at the logs of the last action. do you have any ideas on how to go about this? do we handle all the edge cases? or do we require the user pass in `input_data` instead of `input_size` (only while using `half()` models)?

**Comment by TylerYep** [comment]
Can you clarify what are the edge cases that `half()` is unable to be used?

If `half()` can never be used with CPU, we should raise a clear error message explaining why and how to fix this (e.g. setting `device` correctly).

I don't see an easy way for a user to specify `half()` with `input_size`. For now, I would recommend raising an error saying we do not support half() with `input_size`. In a future change, we can perhaps add a new parameter to `torchinfo.summary()` that lets the user specify the quantization.

File: torchinfo/layer_info.py [hunk 1]
```diff
@@ -55,10 +57,10 @@ def __repr__(self) -> str:
     @staticmethod
     def calculate_size(
         inputs: DETECTED_INPUT_OUTPUT_TYPES, batch_dim: int | None
-    ) -> list[int]:
+    ) -> tuple[list[int], int]:
         """Set input_size or output_size using the model's inputs."""
```
Comment (TylerYep) [inline torchinfo/layer_info.py:61]: Let's add to this comment explaining what the two return values represent

**Comment by jedi2610** [comment]
okay cool i'll go with warning the user cause i feel the same way too

**Comment by jedi2610** [comment]
@TylerYep fixed it. i raise an error if `input_size` is used in a half precision setting. 

few more things:
1. since github actions don't run on a computer that has a gpu/cuda, how do we fix the tests for it to pass while running in gh actions while also checking the summary output (in a local machine). should i add those tests in a separate file or something like that?
2. i should add tests to test if the exception is actually being raised in `exceptions_test.py`, so i'll do that later.

File: tests/torchinfo_test.py [hunk 2]
```diff
@@ -143,9 +144,7 @@ def test_pruning() -> None:
     model = SingleInputNet()
     for module in model.modules():
         if isinstance(module, (torch.nn.Conv2d, torch.nn.Linear)):
-            prune.l1_unstructured(  # type: ignore[no-untyped-call]
-                module, "weight", 0.5
-            )
+            prune.l1_unstructured(module, "weight", 0.5)  # type: ignore[no-untyped-call]
```
Comment (TylerYep) [inline tests/torchinfo_test.py:147]: Revert this line change to fix pre-commit

File: torchinfo/layer_info.py [hunk 3]
```diff
@@ -67,44 +73,49 @@ def nested_list_size(inputs: Sequence[Any]) -> list[int]:
                 or not hasattr(inputs, "__getitem__")
                 or not inputs
             ):
-                return []
+                return [], 0
             if isinstance(inputs[0], dict):
                 return nested_list_size(list(inputs[0].items()))
```
Comment (TylerYep) [inline torchinfo/layer_info.py:78]: We should return something from every branch

**Comment by TylerYep** [comment]
Awesome, it's looking really good. I'll take a final pass this weekend before merging. If you get a chance, could you add a couple tests for the two warnings that were added? To get test coverage on those lines.

**Comment by jedi2610** [comment]
@TylerYep yeah i already have that test on my local machine, let me just commit it once i reach home

**Comment by jedi2610** [comment]
@TylerYep there you go

**Comment by TylerYep** [comment]
Awesome!! Thank you for contributing - I'll get a release out sometime soon with this change.
